src/main.rs

```diff
@@ -159,29 +169,79 @@ fn entry() -> ! {
+    let mut fifo = single_cycle_io.fifo;
+    let mut mc = Multicore::new(&mut peripherals.PSM, &mut peripherals.PPB, &mut fifo);
+    let cores = mc.cores();
+    let core1 = &mut cores[1];
+    core1
+        .spawn(unsafe { &mut CORE1_STACK.mem }, move || {
+            core1_main(serializer)
+        })
+        .unwrap();
+    // Safety: enable interrupt for fifo to receive line render requests.
+    // Transfer ownership of this end of the fifo to the interrupt handler.
     unsafe {
-        NVIC::unmask(Interrupt::DMA_IRQ_0);
+        FIFO = MaybeUninit::new(fifo);
+        NVIC::unmask(Interrupt::SIO_IRQ_PROC0);
     }
-    serializer.wait_fifos_full();
-    serializer.enable();
+    init_display_swapcell();
 
     rom();
     ram();
     ram_x();
     ram_y();
 
-    //unsafe { dbg!(&framebuffer::FRAMEBUFFER_16BPP as *const _) };
-    //unsafe { dbg!(&framebuffer::FRAMEBUFFER_8BPP as *const _) };
-
+    let mut count = 0u32;
     loop {
-        led_pin.toggle().unwrap();
-        delay.delay_ms(250);
+        if count % 15 == 0 {
+            led_pin.toggle().unwrap();
+        }
+        count = count.wrapping_add(1);
+        let (mut rb, mut sb) = start_display_list();
+        let height = 480 / VERTICAL_REPEAT as u32;
+        rb.begin_stripe(height - FONT_HEIGHT);
+        rb.end_stripe();
+        sb.begin_stripe(320 / VERTICAL_REPEAT as u32);
+        sb.solid(92, rgb(0xc0, 0xc0, 0xc0));
+        sb.solid(90, rgb(0xc0, 0xc0, 0));
+        sb.solid(92, rgb(0, 0xc0, 0xc0));
+        sb.solid(92, rgb(0, 0xc0, 0x0));
+        sb.solid(92, rgb(0xc0, 0, 0xc0));
+        sb.solid(90, rgb(0xc0, 0, 0));
+        sb.solid(92, rgb(0, 0, 0xc0));
+        sb.end_stripe();
+        sb.begin_stripe(40 / VERTICAL_REPEAT as u32);
+        sb.solid(92, rgb(0, 0, 0xc0));
+        sb.solid(90, rgb(0x13, 0x13, 0x13));
+        sb.solid(92, rgb(0xc0, 0, 0xc0));
+        sb.solid(92, rgb(0x13, 0x13, 0x13));
+        sb.solid(92, rgb(0, 0xc0, 0xc0));
+        sb.solid(90, rgb(0x13, 0x13, 0x13));
+        sb.solid(92, rgb(0xc0, 0xc0, 0xc0));
+        sb.end_stripe();
+        sb.begin_stripe(120 / VERTICAL_REPEAT as u32 - FONT_HEIGHT);
+        sb.solid(114, rgb(0, 0x21, 0x4c));
+        sb.solid(114, rgb(0xff, 0xff, 0xff));
+        sb.solid(114, rgb(0x32, 0, 0x6a));
+        sb.solid(116, rgb(0x13, 0x13, 0x13));
+        sb.solid(30, rgb(0x09, 0x09, 0x09));
+        sb.solid(30, rgb(0x13, 0x13, 0x13));
+        sb.solid(30, rgb(0x1d, 0x1d, 0x1d));
+        sb.solid(92, rgb(0x13, 0x13, 0x13));
+        sb.end_stripe();
+        rb.begin_stripe(FONT_HEIGHT);
+        let text = format!("Hello pico-dvi-rs, frame {count}");
+        let width = rb.text(&text);
+        let width = width + width % 2;
+        rb.end_stripe();
+        sb.begin_stripe(FONT_HEIGHT);
+        sb.pal_1bpp(width, &BW_PALETTE);
+        sb.solid(640 - width, rgb(0, 0, 0));
+        sb.end_stripe();
+        end_display_list(rb, sb);
```

Comment:
I really like this setup. It has some kinks in its interface (having to deal with two builders simultaneously) but I think, as a building block it is solid. In the future, it would be nice to have the user-facing interface merge the `Renderlist` and `Scanlist` building.

Reply:
It's a good question, and I've thought about it some. The challenge is that the granularity of Scanlist and Renderlist can be different. Above you see that as there's a big vertical stripe in which no rendering is being done by Renderlist, but there are a bunch of stripes in Scanlist. Conversely, for something like rendering tiles and sprites, the intermediate buffer can be just a 4bpp buffer the width of one scan line, with the big vertical stripe on the Scanlist side. But if you want different palettes in different regions of the screen, things become interesting again. I'm not sure whether there is a good unifying abstraction. My gut feeling is that we'll need to play with a few different patterns for generating content.